### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by aljones15**
Adds basic tests for sessionService and session:

```js
  sessionService API
    unauthenticated request
      ��� should get a session with no account
    authenticated request
      ��� should get a session with an account
      ��� should logout a session
      ��� should expire after 1 second
      ��� should refresh on get
  session API
    unauthenticated request
      ��� should create a session
      ��� should get a session with no data
    authenticated request
      ��� should get a session with data
      ��� should logout a session
      ��� should expire after 1 second
      ��� should refresh
      ��� should emit change event on refesh when expired
      ��� should emit change event on end
      ��� should emit change event if authentication passed to refresh

Finished in 12.417 secs / 8.727 secs @ 18:25:26 GMT-0400 (Eastern Daylight Time)

SUMMARY:
��� 14 tests completed
TOTAL: 14 SUCCESS
All tests passed.
```

### CODE REVIEW

File: test/test.config.js [hunk 1]
```diff
@@ -24,3 +26,29 @@ config.mongodb.dropCollections.collections = [];
...
+
+config.express.session.secret = 'NOTASECRET';
+config.express.session.key = 'web-authn-token-test-session';
+config.express.session.prefix = 'web-authn-token-test';
+// test specific settings
+config.express.session.rolling = true;
+config.express.session.resave = true;
+config.express.session.saveUninitialized = false;
+
+// FIXME change this to a common setting
```
Comment (mattcollier) [inline test/test.config.js]: What does this mean?  Can it be resolved?

File: test/web/10-sessionService.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        session = await sessionService.get();
+      } catch(e) {
+        err = e;
+      }
+      should.not.exist(err);
+      should.exist(session);
+      session.should.be.an('object');
+      const keys = Object.keys(session);
+      // an unauthenticated session has no data
+      keys.should.deep.equal([]);
```
Comment (mattcollier) [inline test/web/10-sessionService.js]:
```suggestion
      // an unauthenticated session has no data
      session.should.eql({});
```
`eql` is a shorthand for `deep.equal` 

Please use `foo.should.have.keys(['a', 'b']);` for this sort of assertion elsewhere in this test suite.

File: test/web/10-sessionService.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    });
+  }); // end unauthenticated request
+  describe('authenticated request', () => {
+    let session, account, totp, email, password = null;
```
Comment (mattcollier) [inline test/web/10-sessionService.js]: code stye does not allow defining multiple variables on one line, please separate each def to a separate line.

File: test/web/10-sessionService.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+      session = null;
+      await login({email, password, totp});
+    });
+    afterEach(async function() {
+      await sessionService.logout();
+    });
+    it('should get a session with an account', async () => {
+      let err;
+      try {
+        session = await sessionService.get();
```
Comment (mattcollier) [inline test/web/10-sessionService.js:55]: is the code simpler if `session` is scoped to each `it`?  If one `it` is reusing `session` from a prior `it` that is a pattern to be avoided if possible.

File: test/web/10-sessionService.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      should.not.exist(err);
+      should.exist(session);
+      session.should.be.an('object');
+      let keys = Object.keys(session);
+      // an authenticated session has data
+      keys.should.deep.equal(['account']);
+      session.account.should.be.an('object');
+      session.account.should.have.property('id');
+      session.account.id.should.equal(account.id);
+      for(let i = 0; i < 4; i++) {
```
Comment (mattcollier) [inline test/web/10-sessionService.js]: I think 5 iterations here is more to the point.  Also, please add a comment that you are demonstrating that the session would have expired at 1000ms if the refresh were not working properly.  Just provide the reader more information about the aims of the test and how your are accomplishing them.

File: test/web/20-session.js [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+      should.exist(session);
+      session.should.be.an('object');
+      let keys = Object.keys(session);
+      keys.should.deep.equal(['data', '_service', '_eventTypeListeners']);
+      session.data.should.be.an('object');
+      Object.keys(session.data).should.deep.equal(['account']);
+      session.data.account.should.be.an('object');
+      session.data.account.should.have.property('id');
+      session.data.account.id.should.equal(account.id);
+      for(let i = 0; i < 4; i++) {
```
Comment (mattcollier) [inline test/web/20-session.js]: same comment as above.

File: test/web/20-session.js [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+        Object.keys(session.data).should.deep.equal(['account']);
+        session.data.account.should.be.an('object');
+        session.data.account.should.have.property('id');
+        session.data.account.id.should.equal(account.id);
+        const changeEvent = new Promise(resolve => {
+          session.on('change', ({authentication, oldData, newData}) => {
+            should.not.exist(authentication);
+            should.exist(oldData);
+            should.exist(newData);
+            oldData.should.not.deep.equal(newData);
```
Comment (mattcollier) [inline test/web/20-session.js]: if any of these things throw (the assertion fails), then it should call `reject(err);` on your Promise.  So, all these assertions should be in a try/catch, catch the error and call `reject` with it.

File: test/web/20-session.js [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+      let keys = Object.keys(session);
+      keys.should.deep.equal(['data', '_service', '_eventTypeListeners']);
+      session.data.should.be.an('object');
+      // an authenticated session has data
+      Object.keys(session.data).should.deep.equal(['account']);
+      session.data.account.should.be.an('object');
+      session.data.account.should.have.property('id');
+      session.data.account.id.should.equal(account.id);
+      const changeEvent = new Promise(resolve => {
+        session.on('change', ({authentication, oldData, newData}) => {
```
Comment (mattcollier) [inline test/web/20-session.js:221]: same comment as above.

File: test/web/20-session.js [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        const changeEvent = new Promise(resolve => {
+          authSpy = spy(() => resolve());
+          session.on('change', authSpy);
+        });
+        await session.refresh({authentication: expectedAuth});
+        keys = Object.keys(session);
+        keys.should.deep.equal(['data', '_service', '_eventTypeListeners']);
+        Object.keys(session.data).should.deep.equal(['account']);
+        await changeEvent;
+        authSpy.withArgs({
```
Comment (mattcollier) [inline test/web/20-session.js:272]: nice!

File: test/web/helpers.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+export async function createAccount({
+  email,
+  password,
+  short_name = 'session-test'
+}) {
+  // check to make sure the account does not already exist.
+  const exists = await accountService.exists({email});
+  if(exists) {
+    // if the account exists just return
+    return;
```
Comment (mattcollier) [inline test/web/helpers.js]: I think the better pattern here is to throw a `DuplicateError` and write the caller to handle `DuplicateError`.

This does beg the question, what does `accountService.create` do already?  Maybe this existence check is not needed if `create` already throws a `DuplicateError`.

File: test/web/helpers.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    serviceId: short_name
+  });
+  await tokenService.create({
+    account: account.id,
+    type: 'password',
+    password,
+    authenticationMethod: 'password-test',
+    serviceId: short_name
+  });
+  return {account, email, password, totp};
```
Comment (mattcollier) [inline test/web/helpers.js]: returning values that have just been passed into this function: `email`, `password` is an anti-pattern.  Please only return new data that is generated by this function.

**Review by mattcollier — CHANGES_REQUESTED** [review]
Thanks for working on this.  This is an important feature that we've needed for a long-long time.

File: test/web/helpers.js (hunk 10 above)
Comment (aljones15) [reply test/web/helpers.js]: this is for setting up test data, however it does mean the failure case is not throwing so I will add a `DuplicateError`.

File: test/test.config.js (hunk 1 above)
Comment (aljones15) [reply test/test.config.js]: this is being resolved in the session expiration work that this test suite will help with. So basically in the PR for session expire this FIXME should be removed. This is just a reminder to do it then.

Comment (mattcollier) [reply test/test.config.js]: Please remove this then as it's not actionable by anyone but yourself and create a github issue if you need a reminder to do something.  Generally speaking, adding TODO's FIXME's to code going forward is not really going to be a thing.  It's too difficult to identify and triage FIXMEs and TODOs months after the fact.

Comment (aljones15) [reply test/test.config.js]: Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/68a177b0cbcb2b52ac7eb1f27fe6acbb50cf8f26

Issue added: https://github.com/digitalbazaar/bedrock-web-session/issues/10

File: test/web/10-sessionService.js (hunk 4 above)
Comment (aljones15) [reply test/web/10-sessionService.js:55]: That's a really good idea for this test so I did that: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/68a177b0cbcb2b52ac7eb1f27fe6acbb50cf8f26

File: test/web/20-session.js (hunk 6 above)
Comment (aljones15) [reply test/web/20-session.js]: Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/57c346b2c53b1f3110f15b5356fc0f937a0cbaec

File: test/web/10-sessionService.js (hunk 5 above)
Comment (aljones15) [reply test/web/10-sessionService.js]: Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/57c346b2c53b1f3110f15b5356fc0f937a0cbaec

File: test/web/helpers.js (hunk 11 above)
Comment (aljones15) [reply test/web/helpers.js]: Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/94f6ce22182e0830e69a81e9f20db591ea4337e4

File: test/web/10-sessionService.js (hunk 3 above)
Comment (aljones15) [reply test/web/10-sessionService.js]: Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/94f6ce22182e0830e69a81e9f20db591ea4337e4

File: test/web/20-session.js (hunk 8 above)
Comment (aljones15) [reply test/web/20-session.js:221]: Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/886ccc59a3e6990576c161773af7940f24466e53

File: test/web/20-session.js (hunk 7 above)
Comment (aljones15) [reply test/web/20-session.js]: Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/886ccc59a3e6990576c161773af7940f24466e53

File: test/web/helpers.js (hunk 10 above)
Comment (aljones15) [reply test/web/helpers.js]: Addressed this here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/4e251232ae3d11ad00b1a4a4e74499e85c8084d0

File: test/test.config.js [hunk 12]
```diff
@@ -24,3 +29,28 @@ config.mongodb.dropCollections.collections = [];
...
+
+// this needs to be true in order to set auth stuff
+config['account-http'].autoLoginNewAccounts = true;
+
+config.express.session.secret = 'NOTASECRET';
+config.express.session.key = 'web-authn-token-test-session';
+config.express.session.prefix = 'web-authn-token-test';
+// test specific settings
+config.express.session.rolling = true;
+config.express.session.resave = true;
```
Comment (dlongley) [inline test/test.config.js]: I don't quite understand what's going on here. Are these the defaults? If not, why are they being set here -- don't we want to (also/at least) test against the defaults? If they are the defaults, why are they being set here?

Comment (aljones15) [reply test/test.config.js]: addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/878d7d92c8856c6ea08ec0500d428bca88096309

**Comment by aljones15** [comment]
I ran the last [commit](https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/812baca1c7338f8a3d49c614d3b9e1914dab6d80) 3 times via github actions and did not get the timing error many previous commits got.

**Review by mattcollier — APPROVED** [review]
TY.  Merging.

**Pull request merged**